game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/OffensiveGeneralRetreat.java

```diff
@@ -61,71 +97,95 @@ public void retreatUnits(final IDelegateBridge bridge) {
-        battleState.getGameData(),
-        battleState::getAttackerRetreatTerritories,
-        battleState.isAmphibious());
-  }
+    final CompositeChange change = new CompositeChange();
 
-  private boolean canAttackerRetreatSeaPlanes() {
-    return battleState.getBattleSite().isWater()
-        && battleState.getUnits(BattleState.Side.OFFENSE).stream().anyMatch(Matches.unitIsAir());
+    change.add(retreater.extraRetreatChange(retreatTo, retreatUnits));
+
+    final Map<Retreater.RetreatLocation, Collection<Unit>> retreatingUnitMap =
+        retreater.splitRetreatUnits(retreatUnits);
+
+    retreatSameTerritoryUnits(
+        bridge, retreatingUnitMap.get(Retreater.RetreatLocation.SAME_TERRITORY));
+    change.add(
+        retreatOtherTerritoryUnits(
+            bridge, retreatTo, retreatingUnitMap.get(Retreater.RetreatLocation.OTHER_TERRITORY)));
+
+    bridge.addChange(change);
+
+    if (battleState.getUnits(BattleState.Side.OFFENSE).isEmpty()) {
+      battleActions.endBattle(IBattle.WhoWon.DEFENDER, bridge);
+    } else {
+      bridge.getDisplayChannelBroadcaster().notifyRetreat(battleState.getBattleId(), retreatUnits);
+    }
+
+    bridge
+        .getDisplayChannelBroadcaster()
+        .notifyRetreat(
+            battleState.getAttacker().getName() + retreater.getShortBroadcastSuffix(),
+            battleState.getAttacker().getName() + retreater.getLongBroadcastSuffix(retreatTo),
+            getName(),
+            battleState.getAttacker());
   }
 
-  private boolean canAttackerRetreatPartialAmphib() {
-    if (!Properties.getPartialAmphibiousRetreat(battleState.getGameData())) {
-      return false;
+  private void retreatSameTerritoryUnits(
+      final IDelegateBridge bridge, final Collection<Unit> retreatingUnits) {
+    if (retreatingUnits != null && !retreatingUnits.isEmpty()) {
+      battleState.retreatUnits(BattleState.Side.OFFENSE, retreatingUnits);
+      final String transcriptText = MyFormatter.unitsToText(retreatingUnits) + " retreated";
+      bridge.getHistoryWriter().addChildToEvent(transcriptText, new ArrayList<>(retreatingUnits));
     }
-    // Only include land units when checking for allow amphibious retreat
-    return battleState.getUnits(BattleState.Side.OFFENSE).stream()
-        .filter(Matches.unitIsLand())
-        .anyMatch(Predicate.not(Unit::getWasAmphibious));
   }
 
-  private boolean canAttackerRetreatAmphibPlanes() {
-    final GameData gameData = battleState.getGameData();
-    return (Properties.getWW2V2(gameData)
-            || Properties.getAttackerRetreatPlanes(gameData)
-            || Properties.getPartialAmphibiousRetreat(gameData))
-        && battleState.getUnits(BattleState.Side.OFFENSE).stream().anyMatch(Matches.unitIsAir());
+  private Change retreatOtherTerritoryUnits(
+      final IDelegateBridge bridge,
+      final Territory retreatTo,
+      final Collection<Unit> retreatingUnits) {
+    if (retreatingUnits != null && !retreatingUnits.isEmpty()) {
```

Comment:
ditto on burying control flow in private methods, this should probably be pushed up a layer so it's obvious to a caller that this is a no-op method in certain conditions. Arguably it's a bug whenever a no-op method is invoked (Did the caller know they would get a no-op, if so, why would they call a no-op method? If the caller did not know, is the caller in an unknown state and computing garbage data?)

Reply:
Codeclimate complained about too much complexity in the `retreat` method so I created these helper methods to move the `if` statements away.

I, personally, don't see a problem with calling a no-op method when using abstractions.  In this case, the `Retreater` classes may or may not have these types of retreating instances.  The caller doesn't need to care.  It just needs to ask the `Retreater` for the data (which might be empty or null) and the helper method does the checks.

Reply:
In this case it hurts a bit as you lose context of why retreating units can be null (it can't). It makes it seem like this is called from multiple places. One needs to be careful about void method calls. They are generally poor abstractions. 

For example, what if someone moves where this invoked or adds something in between where the method is called and where these variables are set. 

Another scenario, what if you need an 'else' statement, in this case the abstraction breaks down completely since you are doing two different things. I think it is analgous to this:

```
   doPrinting(Connection c);
}

void doPrinting(Connection c) {
     if(c.connected) {
          sendToFax();
     }
}
```

vs:
```

if(c.connected()) {
  sendToFax();
}
```

And with an if/else branch:

```
   doPrinting(Connection c);
}

private void doPrinting(Connection c) {
     if(c.connected) {
          sendToFax();
     } else {
          sendToCopier();
    }
}
```

In the above, you can see it is just as well to have the if/else statement at the higher level. I think it also comes back to trying to favor having control flow in one place, at a high level, then keep the private methods that are invoked to be functional. 

Last comment, when a no-op method, or a method is carved off with control-flow, it often means the abstraction is cut at a slightly awkward level. The method is not quite control flow, not quite functional. Either the method should do more control flow and invoke other methods, or it should be just the body of the if statement. So  I believe from that it is a code smell.

Reply:
I do think this is more cohesive and actually simpler since you (AFAIK) you don't have to worry about the empty change:

```
    final var retreatingUnits = retreatingUnitMap.get(Retreater.RetreatLocation.OTHER_TERRITORY);
    if (retreatingUnits != null && !retreatingUnits.isEmpty()) {
      change.add(retreatOtherTerritoryUnits(bridge, retreatTo, retreatingUnits));
    }

    bridge.addChange(change);

    if (battleState.getUnits(BattleState.Side.OFFENSE).isEmpty()) {
      battleActions.endBattle(IBattle.WhoWon.DEFENDER, bridge);
    } else {
      bridge.getDisplayChannelBroadcaster().notifyRetreat(battleState.getBattleId(), retreatUnits);
    }
```

Reply:
Even more, the if condition of the private method is duplicated, so the control flow is actually kind loopy to understand. We can simplify by combining, eg:

```
    final Map<Retreater.RetreatLocation, Collection<Unit>> retreatingUnitMap =
        retreater.splitRetreatUnits(retreatUnits);
    final var retreatingUnits = retreatingUnitMap.get(Retreater.RetreatLocation.OTHER_TERRITORY);
    if (retreatingUnits != null && !retreatingUnits.isEmpty()) {
      retreatSameTerritoryUnits(
          bridge, retreatingUnitMap.get(Retreater.RetreatLocation.SAME_TERRITORY));
      change.add(retreatOtherTerritoryUnits(bridge, retreatTo, retreatingUnits));
    }

    bridge.addChange(change);

    if (battleState.getUnits(BattleState.Side.OFFENSE).isEmpty()) {
      battleActions.endBattle(IBattle.WhoWon.DEFENDER, bridge);
    } else {
      bridge.getDisplayChannelBroadcaster().notifyRetreat(battleState.getBattleId(), retreatUnits);
    }
```

Reply:
To fix the complexity of this method, I would do the above, but also make the 'retreat' method actually do a retreat. It violates the "yoda" principle: "Do or do not, there is no try".

Specifically here:
```
    if (retreatTo == null) {
      return;
    }
```

In this case we have a 'retreat' method that is only conditionally doing a retreat.  If the method says 'retreat', I would expect it to do an unconditional retreat.

In this case I would do similar and push the control flow up a level so that this method has a SRP to it and does "retreat". In that case it then becomes the retreat sequence and no longer some of the logic if a retreat should be initiated *and* the actual sequence.

EG:
```
    if (retreater != null) {
      final Collection<Unit> retreatUnits = retreater.getRetreatUnits();
      final Collection<Territory> possibleRetreatSites =
          retreater.getPossibleRetreatSites(retreatUnits);
      final String text = retreater.getQueryText();
      bridge.getDisplayChannelBroadcaster().gotoBattleStep(battleState.getBattleId(), getName());
      final Territory retreatTo =
          battleActions.queryRetreatTerritory(
              battleState, bridge, battleState.getAttacker(), possibleRetreatSites, text);
      if (retreatTo != null) {
        retreat(bridge, retreater, retreatTo);
      }
    }
```

This makes the 'retreatUnits' method more worthwhile, more control flow is pushed up. Now if you get to 'retreat', you actually know that is executed and is not yet more control flow and potentially a no-op.

It makes the retreat method much more cohesive too IMHO:
```
  private void retreat(final IDelegateBridge bridge, final Retreater retreater, final Territory retreatTo) {
    final Collection<Unit> retreatUnits = retreater.getRetreatUnits();

    SoundUtils.playRetreatType(
        battleState.getAttacker(), retreatUnits, retreater.getRetreatType(), bridge);

    final CompositeChange change = new CompositeChange();
   :
   :
```

I would probably go on to rewrite the retreat method as follows:

```
  private void retreat(final IDelegateBridge bridge, final Retreater retreater, final Territory retreatTo) {
    final Collection<Unit> retreatUnits = retreater.getRetreatUnits();

    SoundUtils.playRetreatType(
        battleState.getAttacker(), retreatUnits, retreater.getRetreatType(), bridge);

    bridge.addChange(computeRetreatChange(...));
    
    if (battleState.getUnits(BattleState.Side.OFFENSE).isEmpty()) {
      battleActions.endBattle(IBattle.WhoWon.DEFENDER, bridge);
    } else {
      bridge.getDisplayChannelBroadcaster().notifyRetreat(battleState.getBattleId(), retreatUnits);
    }

    bridge
        .getDisplayChannelBroadcaster()
        .notifyRetreat(
            battleState.getAttacker().getName() + retreater.getShortBroadcastSuffix(),
            battleState.getAttacker().getName() + retreater.getLongBroadcastSuffix(retreatTo),
            getName(),
            battleState.getAttacker());
  }
```

I think I woudl stop there for rewriting that method since you really need to then pivot functionality out of the bridge and I think it hits diminishing returns.. 

IDK if this was overly pedantic for you, hopefully helpful maybe for some ideas, WDYT @trevan ?

Reply:
I've reworked it.